jina/drivers/__init__.py

```diff
@@ -319,20 +323,34 @@ def pre_traverse(_docs, recur_on, context_doc=None,
         else:
             raise ValueError(f'{self.recursion_order}')
 
-        if (self._granularity_start < self._granularity_end) or \
-                (self._granularity_start == self._granularity_end and self._granularity_start > 0):
-            _traverse(docs, 'chunks', None, 'granularity', self._granularity_start, self._granularity_end)
-
-        if (self._adjacency_start < self._adjacency_end) or \
-                (self._adjacency_start == self._adjacency_end and self._adjacency_start > 0):
-
-            for d in docs:
-                # Move to starting depth range
-                # assume that these search docs have a maximum of one chunk per document (common pattern in search)
-                working_doc = d
-                while working_doc.granularity < self._granularity_start:
-                    working_doc = working_doc.chunks[0]
-                _traverse(working_doc.matches, 'matches', None, 'adjacency', self._adjacency_start, self._adjacency_end)
+        if not docs:
+            return
+
+        if 'chunks' in self._recur_on:
+            working_docs = docs
+            while working_docs[0].adjacency < self._adjacency_start:
+                _temp = []
+                for working_doc in working_docs:
+                    _temp.extend(working_doc.matches)
+                working_docs = _temp
+            _traverse(working_docs, 'chunks', None, 'granularity', self._granularity_start, self._granularity_end)
+
+        if 'matches' in self._recur_on:
+            working_docs = docs
+            while working_docs[0].granularity < self._granularity_start:
+                _temp = []
+                for working_doc in working_docs:
+                    _temp.extend(working_doc.chunks)
+                working_docs = _temp
+
+            if 'chunks' in self._recur_on and self._granularity_start == 0 and self._adjacency_start == 0:
```

Comment:
Cant this option be handled before filling the working docs?

Reply:
Great idea. :+1: